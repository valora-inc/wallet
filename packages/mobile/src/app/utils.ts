import { DappRequestOrigin } from 'src/analytics/types'
import { ActiveDapp } from 'src/app/reducers'

// Assume that if we have an active dapp, any WC or DappKit request comes from the in-app webview.
// Note that this may be incorrect if the user interacts with another dapp while the dapp webview is open.
// I thought about comparing the request url to the active dapp url, but that could fail too if the dapp url we have is different.
// For instance the dapp-list could contain a "example.com", while the request is "dapp.something.com".
export function getDappRequestOrigin(activeDapp: ActiveDapp | null) {
  if (activeDapp) {
    return DappRequestOrigin.InAppWebView
  }

  return DappRequestOrigin.External
}
