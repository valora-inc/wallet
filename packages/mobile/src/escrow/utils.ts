import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'
import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToE164NumberSelector } from 'src/identity/selectors'
import { ContactRecipient } from 'src/recipients/recipient'
import useSelector from 'src/redux/useSelector'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}

export const useEscrowPaymentRecipient = (payment: EscrowedPayment): ContactRecipient => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToE164Number = useSelector(identifierToE164NumberSelector)
  const recipientCache = useSelector((state) => state.recipients.phoneRecipientCache)
  const phoneNumber = identifierToE164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber] ?? {
    name: phoneNumber,
    e164PhoneNumber: phoneNumber,
    contactId: '',
  }

  return recipient
}
