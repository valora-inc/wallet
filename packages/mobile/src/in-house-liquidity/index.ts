import { FinclusiveKycStatus } from 'src/account/reducer'
import networkConfig from 'src/geth/networkConfig'

interface RequiredParams {
  walletAddress: string
}

/**
 * get the status of a users finclusive compliance check aka their KYC status
 *
 *
 * @param {params.walletAddress} walletAddress
 * @param {params.dekPrivate} dekPrivate private data encryption key
 * @returns {FinclusiveKycStatus} the users current status
 */
export const getFinclusiveComplianceStatus = async ({
  walletAddress,
}: RequiredParams): Promise<FinclusiveKycStatus> => {
  const response = await signAndFetch({
    path: `/account/${encodeURIComponent(walletAddress)}/compliance-check-status`,
    walletAddress,
    requestOptions: {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    },
  })
  if (!response.ok) {
    throw new Error(
      `IHL GET /account/:accountAddress/compliance-check-status failure status ${response.status}`
    )
  }
  const { complianceCheckStatus } = await response.json()
  return complianceCheckStatus
}

type DeleteFinclusiveBankAccountParams = RequiredParams & {
  id: number
}

/**
 * get a fiat bank account from finclusive
 *
 * @param {params.walletAddress} walletAddress
 * @param {params.dekPrivate} dekPrivate private data encryption key
 */
export const deleteFinclusiveBankAccount = async ({
  walletAddress,
  id,
}: DeleteFinclusiveBankAccountParams): Promise<void> => {
  const body = {
    accountAddress: walletAddress,
    accountId: id,
  }
  const response = await signAndFetch({
    path: `/account/bank-account?accountAddress=${encodeURIComponent(walletAddress)}`,
    walletAddress,
    requestOptions: {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    },
  })
  if (!response.ok) {
    throw new Error(`IHL DELETE /account/bank-account failure status ${response.status}`)
  }
}

export interface BankAccount {
  id: number
  accountName: string
  accountType: string
  accountNumberTruncated: string
  institutionName: string
  institutionLogo?: string
}

/**
 * get a users fiat bank accounts from finclusive
 *
 *
 * @param {params.walletAddress} walletAddress
 * @param {params.dekPrivate} dekPrivate private data encryption key
 * @returns {BankAccounts} List of bank accounts that the user has linked
 */
export const getFinclusiveBankAccounts = async ({
  walletAddress,
}: RequiredParams): Promise<BankAccount[]> => {
  const response = await signAndFetch({
    path: `/account/bank-account?accountAddress=${encodeURIComponent(walletAddress)}`,
    walletAddress,
    requestOptions: {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    },
  })
  if (!response.ok) {
    throw new Error(`IHL GET /account/bank-account failure status ${response.status}`)
  }
  const { bankAccounts } = await response.json()
  return bankAccounts
}

type CreateFinclusiveBankAccountParams = RequiredParams & {
  plaidAccessToken: string
}

/**
 * Create a fiat bank account with finclusive
 *
 *
 * @param {params.walletAddress} walletAddress
 * @param {params.plaidAccessToken} plaidAccessToken plaid long term access token
 */
export const createFinclusiveBankAccount = async ({
  walletAddress,
  plaidAccessToken,
}: CreateFinclusiveBankAccountParams): Promise<void> => {
  const body = {
    accountAddress: walletAddress,
    plaidAccessToken,
  }
  const response = await signAndFetch({
    path: '/account/bank-account',
    walletAddress,
    requestOptions: {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    },
  })
  if (!response.ok) {
    throw new Error(`IHL POST /account/bank-account failure status ${response.status}`)
  }
}

type ExchangePlaidAccessTokenParams = RequiredParams & {
  publicToken: string
}

/**
 * Exchange a plaid plublic token for a long-term plaid access token
 *
 *
 * @param {params.walletAddress} walletAddress
 * @param {params.publicToken} publicToken plaid public token
 * @returns {accessToken} string accesstoken from plaid
 */
export const exchangePlaidAccessToken = async ({
  walletAddress,
  publicToken,
}: ExchangePlaidAccessTokenParams): Promise<string> => {
  const body = {
    publicToken,
    accountAddress: walletAddress,
  }
  const response = await signAndFetch({
    path: '/plaid/access-token/exchange',
    walletAddress,

    requestOptions: {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    },
  })
  if (!response.ok) {
    throw new Error(`IHL /plaid/access-token/exchange failure status ${response.status}`)
  }
  const { accessToken } = await response.json()
  return accessToken
}

type CreateLinkTokenParams = RequiredParams & {
  isAndroid: boolean
  language: string
  accessToken?: string
  phoneNumber: string
}

/**
 * Create a new Plaid Link Token by calling IHL
 *
 *
 * @param {params.walletAddress} walletAddress
 * @param {params.isAndroid} isAndroid
 * @param {params.language} language the users current language
 * @param {params.accessToken} accessToken optional access token used for editing existing items
 * @param {params.phoneNumber} phoneNumber users verified phone number
 * @returns {linkToken} the link token from the plaid backend
 */
export const createLinkToken = async ({
  walletAddress,

  isAndroid,
  language,
  accessToken,
  phoneNumber,
}: CreateLinkTokenParams): Promise<string> => {
  const body = {
    accountAddress: walletAddress,
    isAndroid,
    language,
    accessToken,
    phoneNumber,
  }
  const response = await signAndFetch({
    path: '/plaid/link-token/create',
    walletAddress,

    requestOptions: {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    },
  })
  if (!response.ok) {
    throw new Error(`IHL /plaid/link-token/create failure status ${response.status}`)
  }
  const { linkToken } = await response.json()
  return linkToken
}

/**
 * Create a Persona account for the given accountMTWAddress
 *
 *
 * @param {params.walletAddress} walletAddress
 */
export const createPersonaAccount = async ({ walletAddress }: RequiredParams): Promise<void> => {
  const body = { accountAddress: walletAddress }
  const response = await signAndFetch({
    path: '/persona/account/create',
    walletAddress,
    requestOptions: {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    },
  })
  if (response.status !== 201 && response.status !== 409) {
    throw new Error(`IHL /persona/account/create failure status ${response.status}`)
  }
}

interface SignAndFetchParams {
  path: string
  walletAddress: string
  requestOptions: RequestInit
}

/**
 * A fetch wrapper that adds in the signature needed for IHL authorization
 *
 *
 * @param {params.path} string like /persona/get/foo
 * @param {params.requestOptions} requestOptions all the normal fetch options
 * @returns {Response} response object from the fetch call
 */
export const signAndFetch = async ({
  path,
  requestOptions,
}: SignAndFetchParams): Promise<Response> => {
  return fetch(`${networkConfig.inHouseLiquidityURL}${path}`, {
    ...requestOptions,
    headers: {
      ...requestOptions.headers,
      // Authorization: authHeader, // todo add this once auth scheme is refactored
    },
  })
}

// just checks that the wallet address is non null for now. keeping this since we may want it to do more once
//  the new auth scheme is figured out (like checking if a wallet is unlocked or a token is truthy)
export const verifyWalletAddress = ({
  walletAddress,
}: {
  walletAddress: string | null
}): RequiredParams => {
  if (!walletAddress) {
    throw new Error('Cannot call IHL because walletAddress is null')
  }

  return {
    walletAddress: walletAddress,
  }
}
