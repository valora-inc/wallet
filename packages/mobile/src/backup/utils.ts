import { MnemonicLanguages, normalizeMnemonic } from '@celo/utils/lib/account'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import { useDispatch, useSelector } from 'react-redux'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

export const DAYS_TO_BACKUP = 1
export const DAYS_TO_DELAY = 1 / 24 // 1 hour delay
export const MNEMONIC_SPLITTER = 'celo'

export const MNEMONIC_STORAGE_KEY = 'mnemonic'

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Account Key',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function formatBackupPhraseOnSubmit(phrase: string) {
  return normalizeMnemonic(phrase)
}

function isValidMnemonic(phrase: string, length: number) {
  return !!phrase && countMnemonicWords(formatBackupPhraseOnEdit(phrase)) === length
}

export function isValidBackupPhrase(phrase: string) {
  return isValidMnemonic(phrase, 24)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
