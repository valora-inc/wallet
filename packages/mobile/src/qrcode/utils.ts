import * as RNFS from 'react-native-fs'
import Share from 'react-native-share'
import { call, put, select } from 'redux-saga/effects'
import { showError, showMessage } from 'src/alert/actions'
import { SendEvents } from 'src/analytics/Events'
import { SendOrigin, WalletConnectPairingOrigin } from 'src/analytics/types'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { walletConnectEnabledSelector } from 'src/app/selectors'
import { validateRecipientAddressSuccess } from 'src/identity/actions'
import { E164NumberToAddressType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { UriData, uriDataFromUrl } from 'src/qrcode/schema'
import {
  getRecipientFromAddress,
  recipientHasNumber,
  RecipientInfo,
} from 'src/recipients/recipient'
import { QrCode, SVG } from 'src/send/actions'
import { TransactionDataInput } from 'src/send/SendAmount'
import { handleSendPaymentData } from 'src/send/utils'
import Logger from 'src/utils/Logger'
import { initialiseWalletConnect } from 'src/walletConnect/saga'

export enum BarcodeTypes {
  QR_CODE = 'QR_CODE',
}

const TAG = 'QR/utils'

const QRFileName = '/celo-qr.png'

export async function shareSVGImage(svg: SVG) {
  if (!svg) {
    return
  }
  const data = await new Promise<string>((resolve, reject) => {
    svg.toDataURL((dataURL: string | undefined) => {
      if (dataURL) {
        resolve(dataURL)
      } else {
        // Not supposed to happen, but throw in case it does :)
        reject(new Error('Got invalid SVG data'))
      }
    })
  })

  const path = RNFS.DocumentDirectoryPath + QRFileName
  await RNFS.writeFile(path, data, 'base64')
  return Share.open({
    url: 'file://' + path,
    type: 'image/png',
    failOnCancel: false, // don't throw if user cancels share
  })
}

function* handleSecureSend(
  address: string,
  e164NumberToAddress: E164NumberToAddressType,
  secureSendTxData: TransactionDataInput,
  requesterAddress?: string
) {
  if (!recipientHasNumber(secureSendTxData.recipient)) {
    throw Error('Invalid recipient type for Secure Send, has no mobile number')
  }

  const userScannedAddress = address.toLowerCase()
  const { e164PhoneNumber } = secureSendTxData.recipient
  const possibleReceivingAddresses = e164NumberToAddress[e164PhoneNumber]
  // This should never happen. Secure Send is triggered when there are
  // multiple addresses for a given phone number
  if (!possibleReceivingAddresses) {
    throw Error("No addresses associated with recipient's phone number")
  }

  // Need to add the requester address to the option set in the event
  // a request is coming from an unverified account
  if (requesterAddress && !possibleReceivingAddresses.includes(requesterAddress)) {
    possibleReceivingAddresses.push(requesterAddress)
  }
  const possibleReceivingAddressesFormatted = possibleReceivingAddresses.map((addr) =>
    addr.toLowerCase()
  )
  if (!possibleReceivingAddressesFormatted.includes(userScannedAddress)) {
    const error = ErrorMessages.QR_FAILED_INVALID_RECIPIENT
    ValoraAnalytics.track(SendEvents.send_secure_incorrect, {
      confirmByScan: true,
      error,
    })
    yield put(showMessage(error))
    return false
  }

  ValoraAnalytics.track(SendEvents.send_secure_complete, { confirmByScan: true })
  yield put(validateRecipientAddressSuccess(e164PhoneNumber, userScannedAddress))
  return true
}

export function* handleBarcode(
  barcode: QrCode,
  e164NumberToAddress: E164NumberToAddressType,
  recipientInfo: RecipientInfo,
  secureSendTxData?: TransactionDataInput,
  isOutgoingPaymentRequest?: boolean,
  requesterAddress?: string
) {
  const walletConnectEnabled: boolean = yield select(walletConnectEnabledSelector)
  if (barcode.data.startsWith('wc:') && walletConnectEnabled) {
    navigate(Screens.WalletConnectLoading, { origin: WalletConnectPairingOrigin.Scan })
    yield call(initialiseWalletConnect, barcode.data, WalletConnectPairingOrigin.Scan)
    return
  }

  let qrData: UriData
  try {
    qrData = uriDataFromUrl(barcode.data)
  } catch (e) {
    yield put(showError(ErrorMessages.QR_FAILED_INVALID_ADDRESS))
    Logger.error(TAG, 'qr scan failed', e)
    return
  }

  if (secureSendTxData) {
    const success: boolean = yield call(
      handleSecureSend,
      qrData.address,
      e164NumberToAddress,
      secureSendTxData,
      requesterAddress
    )
    if (!success) {
      return
    }

    if (isOutgoingPaymentRequest) {
      navigate(Screens.PaymentRequestConfirmation, {
        transactionData: secureSendTxData,
        addressJustValidated: true,
      })
    } else {
      navigate(Screens.SendConfirmation, {
        transactionData: secureSendTxData,
        addressJustValidated: true,
        origin: SendOrigin.AppSendFlow,
      })
    }

    return
  }

  const cachedRecipient = getRecipientFromAddress(qrData.address, recipientInfo)

  yield call(handleSendPaymentData, qrData, cachedRecipient, isOutgoingPaymentRequest, true)
}
