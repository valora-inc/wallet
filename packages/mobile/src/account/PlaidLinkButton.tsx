import * as React from 'react'
import Button, { BtnSizes, BtnTypes } from '@celo/react-components/components/Button'
import { openLink } from 'react-native-plaid-link-sdk'
import { createLinkToken } from 'src/in-house-liquidity'
import { mtwAddressSelector, walletAddressSelector } from 'src/web3/selectors'
import { currentLanguageSelector } from 'src/i18n/selectors'
import { e164NumberSelector } from 'src/account/selectors'
import Logger from 'src/utils/Logger'
import { useTranslation } from 'react-i18next'

import { useDispatch, useSelector } from 'react-redux'
import { Platform, StyleSheet } from 'react-native'
import { showError } from 'src/alert/actions'
import { ErrorMessages } from 'src/app/ErrorMessages'

const TAG = 'PLAID'

const PlaidLinkButton = ({ disabled }: { disabled: boolean }) => {
  const dispatch = useDispatch()
  const { t } = useTranslation()

  const accountMTWAddress = useSelector(mtwAddressSelector)
  const walletAddress = useSelector(walletAddressSelector)
  const locale = useSelector(currentLanguageSelector) || ''
  const phoneNumber = useSelector(e164NumberSelector) || ''
  const isAndroid = Platform.OS === 'android'

  const onPress = async () => {
    if (!accountMTWAddress) {
      Logger.error(TAG, "Can't render Plaid because accountMTWAddress is null")
      return
    }
    if (!walletAddress) {
      Logger.error(TAG, "Can't render Plaid because walletAddress is null")
      return
    }
    const IHLResponse = await createLinkToken({
      accountMTWAddress,
      walletAddress,
      isAndroid,
      language: locale.split('-')[0], // ex: just en, not en-US
      phoneNumber,
    })
    if (!IHLResponse.ok) {
      dispatch(showError(ErrorMessages.PLAID_CREATE_LINK_TOKEN_FAIL))
      return
    }
    const { linkToken } = await IHLResponse.json()
    return openLink({
      tokenConfig: { token: linkToken },
      onSuccess: ({ publicToken, metadata }) => {
        // TODO(wallet#1448): call the POST /plaid/access-token/exchange IHL endpoint.
        // TODO(wallet#1448): create a counter-party with finclusive using the POST /account/counter-party endpoint
        // TODO(wallet#1449): redirect to Bank Account List Page
      },
      onExit: () => {
        // TODO(wallet#1447): handle errors from onExit
      },
    })
  }
  return (
    <Button
      style={styles.button}
      onPress={onPress}
      text={t('linkBankAccountScreen.stepTwo.cta')}
      type={BtnTypes.SECONDARY}
      size={BtnSizes.MEDIUM}
      testID="PlaidLinkButton"
      disabled={disabled || !accountMTWAddress || !walletAddress}
    />
  )
}

const styles = StyleSheet.create({
  button: {
    marginTop: 48,
  },
})

export default PlaidLinkButton
