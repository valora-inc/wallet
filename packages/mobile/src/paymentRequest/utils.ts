import { hexToBuffer } from '@celo/utils/lib/address'
import { decryptComment, encryptComment } from '@celo/utils/lib/commentEncryption'
import { isE164Number } from '@celo/utils/lib/phoneNumbers'
import { call } from 'redux-saga/effects'
import { MAX_COMMENT_LENGTH } from 'src/config'
import { features } from 'src/flags'
import i18n from 'src/i18n'
import { PaymentRequest, WriteablePaymentRequest } from 'src/paymentRequest/types'
import Logger from 'src/utils/Logger'
import { doFetchDataEncryptionKey } from 'src/web3/dataEncryptionKey'

const TAG = 'paymentRequest/utils'

// Encrypt sensitive data in the payment request using the recipient and sender DEK
export function* encryptPaymentRequest(paymentRequest: WriteablePaymentRequest) {
  Logger.debug(`${TAG}@encryptPaymentRequest`, 'Encrypting payment request')

  const fromKey: Buffer | null = yield call(
    doFetchDataEncryptionKey,
    paymentRequest.requesterAddress
  )
  if (!fromKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No sender key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const toKey: Buffer | null = yield call(doFetchDataEncryptionKey, paymentRequest.requesteeAddress)
  if (!toKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No recipient key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const encryptedPaymentRequest: WriteablePaymentRequest = {
    ...paymentRequest,
  }

  if (paymentRequest.requesterE164Number) {
    // Using the same util as for comment encryption to encrypt the phone number
    // TODO: Consider renaming this util for clarity
    const { comment: encryptedE164Number, success } = encryptComment(
      paymentRequest.requesterE164Number,
      toKey,
      fromKey
    )

    // We intentionally exclude the phone number if we can't encrypt it
    // The request still contains an address which may still map to a contact
    // If the recipient has seen it before.
    encryptedPaymentRequest.requesterE164Number = success ? encryptedE164Number : undefined
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: encryptedComment, success } = encryptComment(comment, toKey, fromKey)
    encryptedPaymentRequest.comment = success ? encryptedComment : comment
  }

  return encryptedPaymentRequest
}

// Decrypt sensitive data in the payment request using the user's DEK
export function decryptPaymentRequest(
  paymentRequest: PaymentRequest,
  dataEncryptionKey: string | null,
  isOutgoingRequest: boolean
) {
  Logger.debug(`${TAG}@decryptPaymentRequest`, 'Decrypting payment request')

  if (!dataEncryptionKey) {
    Logger.error(`${TAG}@decryptPaymentRequest`, 'Missing DEK, should never happen.')
    return paymentRequest
  }
  const dekBuffer = hexToBuffer(dataEncryptionKey)

  const decryptedPaymentRequest: PaymentRequest = {
    ...paymentRequest,
  }

  const requesterE164Number = paymentRequest.requesterE164Number
  if (requesterE164Number) {
    const { comment: decryptedRequesterE164Number, success } = decryptComment(
      requesterE164Number,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.requesterE164Number = decryptedRequesterE164Number
    } else if (isE164Number(requesterE164Number)) {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Decrypting requesterE164Number failed, using raw number'
      )
      decryptedPaymentRequest.requesterE164Number = requesterE164Number
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'requesterE164Number appears to be ciphertext, excluding it'
      )
      decryptedPaymentRequest.requesterE164Number = undefined
    }
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: decryptedComment, success } = decryptComment(
      comment,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.comment = decryptedComment
    } else if (comment.length <= MAX_COMMENT_LENGTH) {
      Logger.warn(`${TAG}@decryptPaymentRequest`, 'Decrypting comment failed, using raw comment')
      decryptedPaymentRequest.comment = comment
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Comment appears to be ciphertext, hiding comment'
      )
      decryptedPaymentRequest.comment = i18n.t('commentUnavailable')
    }
  }

  return decryptedPaymentRequest
}

// For cases when the request can't be encrypted, remove sensitive PII
function sanitizePaymentRequest(paymentRequest: WriteablePaymentRequest): WriteablePaymentRequest {
  return {
    ...paymentRequest,
    requesterE164Number: undefined,
  }
}
