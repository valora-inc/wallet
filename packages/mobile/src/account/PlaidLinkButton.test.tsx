import { fireEvent, render, waitFor } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { openLink } from 'react-native-plaid-link-sdk'
import { Provider } from 'react-redux'
import PlaidLinkButton from 'src/account/PlaidLinkButton'
import { createMockStore } from 'test/utils'
import { mockAccount, mockAccount2, mockPrivateDEK } from 'test/values'
import { createLinkToken } from 'src/in-house-liquidity'

jest.mock('react-native-plaid-link-sdk', () => ({
  openLink: jest.fn(),
}))
const mockSuccessResponse = new Response(JSON.stringify({ linkToken: 'foo' }), { status: 200 })
const mockFailResponse = new Response(null, { status: 404 })

jest.mock('src/in-house-liquidity', () => ({
  createLinkToken: jest.fn(({ accountMTWAddress }) => {
    if (accountMTWAddress === 'bad-account') {
      return mockFailResponse
    }
    return mockSuccessResponse
  }),
}))

const MOCK_PHONE_NUMBER = '+18487623478'

describe('PlaidLinkButton', () => {
  const store = createMockStore({
    web3: {
      mtwAddress: mockAccount,
      account: mockAccount2,
      dataEncryptionKey: mockPrivateDEK,
    },
    i18n: {
      language: 'en-US',
    },
    account: {
      e164PhoneNumber: MOCK_PHONE_NUMBER,
    },
  })

  beforeEach(() => {
    jest.useRealTimers()
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <PlaidLinkButton disabled={false} />
      </Provider>
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('calls IHL and openLink when pressed', async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <PlaidLinkButton disabled={false} />
      </Provider>
    )
    expect(getByTestId('PlaidLinkButton')).toBeEnabled()

    fireEvent.press(getByTestId('PlaidLinkButton'))
    await waitFor(() =>
      expect(createLinkToken).toHaveBeenCalledWith({
        accountMTWAddress: mockAccount,
        isAndroid: true,
        language: 'en',
        phoneNumber: MOCK_PHONE_NUMBER,
        dekPrivate: mockPrivateDEK,
      })
    )

    expect(openLink).toHaveBeenCalledWith({
      tokenConfig: {
        token: 'foo',
      },
      onExit: expect.anything(),
      onSuccess: expect.anything(),
    })
  })

  it('does not call openLink if IHL returns an error', async () => {
    const store = createMockStore({
      web3: {
        mtwAddress: 'bad-account',
        account: mockAccount2,
        dataEncryptionKey: mockPrivateDEK,
      },
      i18n: {
        language: 'en-US',
      },
      account: {
        e164PhoneNumber: MOCK_PHONE_NUMBER,
      },
    })
    const { getByTestId } = render(
      <Provider store={store}>
        <PlaidLinkButton disabled={false} />
      </Provider>
    )
    expect(getByTestId('PlaidLinkButton')).toBeEnabled()

    fireEvent.press(getByTestId('PlaidLinkButton'))
    await waitFor(() => expect(createLinkToken).toHaveBeenCalled())
    expect(openLink).not.toHaveBeenCalled()
  })
})
