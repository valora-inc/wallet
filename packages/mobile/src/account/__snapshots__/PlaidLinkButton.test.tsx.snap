// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaidLinkButton renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      Object {
        "marginTop": 48,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 100,
        },
      ]
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "rippleRadius": undefined,
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F1F0EB",
          "height": 48,
          "justifyContent": "center",
          "minWidth": 120,
          "opacity": undefined,
          "paddingHorizontal": 24,
          "paddingVertical": 5,
        }
      }
      testID="PlaidLinkButton"
    >
      <Text
        maxFontSizeMultiplier={1}
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 16,
            "lineHeight": 22,
          }
        }
      >
        linkBankAccountScreen.stepTwo.cta
      </Text>
    </View>
  </View>
</View>
`;
